### PR TITLE
[MGDOBR-679] Add deployer bot

### PR DESCRIPTION
**Pull request by r00ta**
[MGDOBR-679](https://issues.redhat.com/browse/MGDOBR-679)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] The layers in the `kustomize` folder have been updated accordingly.
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

**Comment by manstis** [comment]
Is it worth commenting that the original PR needs to presumably be a single commit (I.e squashed and merged)?

**Comment by r00ta** [comment]
> Is it worth commenting that the original PR needs to presumably be a single commit (I.e squashed and merged)?

Yes usually we use `squash and merge` (I hope that everybody is using it, but let me check). In any case it should work also with the other methods, but of course I have to check. 

But we should enforce `squash and merge` in any case

**Review by danielezonca — APPROVED** [review]
LGTM, minor comments

File: .github/workflows/deploy-bot.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+              repo: context.repo.repo,
+              pull_number: context.issue.number
+            }
+            core.info(`Getting PR #${request.pull_number} from ${request.owner}/${request.repo}`)
+            try {
+              const result = await github.pulls.get(request)
+              return result.data
+            } catch (err) {
+              core.setFailed(`Request failed with error ${err}`)
+            }
```
[the remark this answers is not in the document]
Comment (r00ta) [reply .github/workflows/deploy-bot.yml:35]: No

File: .github/workflows/deploy-bot.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        uses: actions-cool/check-user-permission@v2.0.1
+        with:
+          username: ${{ github.actor }}
+          require: developer
+
+      - name: Rebase and deploy
+        id: rebase
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          AUTHOR: ${{ github.actor }}
```
[the remark this answers is not in the document]
Comment (r00ta) [reply .github/workflows/deploy-bot.yml:46]: It's the user that added the comment

File: .github/workflows/deploy-bot.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - uses: actions/github-script@v3
+        id: get-pr
+        with:
+          script: |
+            const request = {
+              owner: context.repo.owner,
+              repo: context.repo.repo,
```
[the remark this answers is not in the document]
Comment (r00ta) [reply .github/workflows/deploy-bot.yml:26]: Nope, I need somebody from the team to try that. Keep you posted

[the remark this answers is not in the document]
Comment (r00ta) [reply .github/workflows/deploy-bot.yml:45]: tested with @lucamolteni , see https://github.com/r00ta/gitops-test/pull/13

**Review by lucamolteni — APPROVED** [review]

File: .github/scripts/rebase-and-deploy-to-env.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+if [[ "$TARGET_BRANCH" == "stable" ]]; then
+  if [ $(git branch --contains $SHA_COMMIT | grep -c "dev") -eq 0 ]; then
+    printf "\U274C In order to deploy to stable branch the feature must be on dev branch first. Please deploy it there first, check that everything is running fine and then promote it to stable."
+    exit 0
+  fi
+fi
+
+# If the commit is already on TARGET_BRANCH, the feature has been already merged and deployed.
+if [ $(git branch --contains $SHA_COMMIT | grep -c "$TARGET_BRANCH") -ne 0 ]; then
+  printf "\U274C $SHA_COMMIT is already on $TARGET_BRANCH branch!"
```
[the remark this answers is not in the document]
Comment (r00ta) [reply .github/scripts/rebase-and-deploy-to-env.sh:34]: Makes sense

[the remark this answers is not in the document]
Comment (r00ta) [reply .github/scripts/rebase-and-deploy-to-env.sh]: Done

**Review by kostola — APPROVED** [review]
LGTM :+1:

**Review by sutaakar — APPROVED** [review]

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -30,5 +30,7 @@ How to trigger pipelines and use the bots:
 
 - <b>Rebase the pull request</b>  
   Comment with: `/rebase`.
+- <b>Deploy BOT</b>
+  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.
```
[the remark this answers is not in the document]
Comment (r00ta) [reply .github/pull_request_template.md:34]: @robpblake there are only two cases: 
1) The pull request triggers the build of *at least* one image -> use the bot in the `update kustomization images` 
2) The pull request does not trigger any build of the images -> use the bot directly in the pull request. (this is the case for integration tests and kustomize configuration only PR for example)

With the time we can improve the bot to guide the developer in the deployment process. But for the time being @manstis @kostola @lucamolteni @danielezonca @sutaakar @vaibhavjainwiz ^^ is the strategy to use the bot. I'll send an email when this gets merged with the instructions.

**Review by robpblake — APPROVED** [review]

**Pull request merged**